File: dotnet/ADP.VINDecode/VIN.cs
using System;

namespace ShiftSoftware.ADP.VINDecode
{
    public class VIN
    {
        private readonly string vin;
        public string WMI { get; }
        public string VDS { get; }
        public string CD { get; }
        public string VIS { get; }

        private VIN(string vin)
        {
            this.vin = vin;
            WMI = this.vin.Substring(0, 3);
            VDS = this.vin.Substring(3, 6);
            CD = this.vin.Substring(9, 1);
            VIS = this.vin.Substring(9, 8);
        }

        private VIN(ReadOnlySpan<char> vin)
        {
            this.vin = vin.ToString().ToUpperInvariant();
            WMI = vin.Slice(0, 3).ToString();
            VDS = vin.Slice(3, 6).ToString();
            CD = vin.Slice(9, 1).ToString();
            VIS = vin.Slice(9, 8).ToString();
        }

        /// <summary>
        /// Tries to parse a VIN. Returns true if successful, otherwise false.
        /// </summary>
        public static bool TryParse(string vin, out VIN result)
        {
            result = null;

            vin = vin.ToUpperInvariant();

            if (!Validator.ValidateCheckDigit(vin))
                return false;

            result = new VIN(vin);

            return true;
        }

        /// <summary>
        /// Tries to parse a VIN. Returns true if successful, otherwise false.
        /// </summary>
        public static bool TryParse(ReadOnlySpan<char> vin, out VIN result)
        {
            Span<char> vinUpper = stackalloc char[17]; // Allocate stack memory for uppercase conversion

            if (vin.Length != 17)
            {
                result = null;
                return false;
            }

            for (int i = 0; i < 17; i++)
            {
                char c = vin[i];

                if (c >= 'a' && c <= 'z') // Convert lowercase to uppercase
                    c = (char)(c - 32);

                vinUpper[i] = c;
            }

            if (!Validator.ValidateCheckDigit(vinUpper))
            {
                result = null;
                return false;
            }

            result = new VIN(vinUpper); // Ensure VIN constructor accepts ReadOnlySpan<char>

            return true;
        }

        /// <summary>
        /// Parses a VIN. Throws an exception if invalid.
        /// </summary>
        public static VIN Parse(string vin)
        {
            vin = vin.ToUpperInvariant();

            if (!Validator.ValidateCheckDigit(vin))
                throw new ArgumentException("Invalid VIN", nameof(vin));

            return new VIN(vin);
        }

        public override string ToString() => vin;
        public override bool Equals(object obj) => obj is VIN other && vin.Equals(other.vin, StringComparison.OrdinalIgnoreCase);
        public override int GetHashCode() => vin.ToUpperInvariant().GetHashCode();

        public static implicit operator string(VIN vin) => vin.ToString();
    }
}

File: android/lib/src/main/kotlin/software/shift/adp/vindecode/VIN.kt
package software.shift.adp.vindecode

class VIN private constructor(private val vin: String) {

    val wmi: String
    val vds: String
    val cd: String
    val vis: String

    init {
        wmi = vin.substring(0, 3)
        vds = vin.substring(3, 9)
        cd = vin.substring(9, 10)
        vis = vin.substring(9, 17)
    }

    companion object {
        /**
         * Tries to parse a VIN. Returns VIN if successful, otherwise null.
         */
        fun tryParseVIN(vin: String): VIN? {
            val normalizedVin = vin.uppercase()
            if (!Validator.validateCheckDigit(normalizedVin)) {
                return null
            }
            return  VIN(normalizedVin)
        }

        /**
         * Tries to parse a VIN. Returns VIN if successful, otherwise null.
         */
        fun tryParseVIN(vin: CharArray):  VIN? {
            if (vin.size != 17) return null

            val vinUpper = CharArray(17) {
                val c = vin[it]
                if (c in 'a'..'z') (c - 32).toChar() else c
            }

            if (!Validator.validateCheckDigit(vinUpper)) {
                return null
            }

            return VIN(String(vinUpper))
        }

        /**
         * Parses a VIN. Throws an exception if invalid.
         */
        fun parseVIN(vin: String): VIN {
            val normalizedVin = vin.uppercase()
            if (!Validator.validateCheckDigit(normalizedVin)) {
                throw IllegalArgumentException("Invalid VIN")
            }
            return VIN(normalizedVin)
        }
    }

    override fun toString(): String = vin

    override fun equals(other: Any?): Boolean {
        return other is VIN && this.vin.equals(other.vin, ignoreCase = true)
    }

    override fun hashCode(): Int {
        return vin.uppercase().hashCode()
    }
}
